### PR TITLE
Multimic preferences@aferr

### PR DESCRIPTION
**Pull request by aferr**
This way of writing multimic covers the PRD and variant1 where not all members of the room have their privacy preferences met. This works by having all purchasers of a mobile device within this policy framework agree to a "mechanized EULA" in the form of a policy principal. Users give preferences to this principal, but the policy principal may not always meet their preferences. This uses a pattern similar to [this one](https://github.com/google-research/raksha/blob/main/rust/tools/authorization-logic/raksha_examples/policy_patterns/NonUnanimousDecisions), and there might be other ways of mapping this "policy design pattern" to reality than "an EULA".

This compiles and runs as expected. To run this:
```
cd raksha/rust/tools/authorization-logic &&
cargo cargo run multimic-preferences.authlogic $RAKSHA/src/analysis/souffle/examples/multimic tmp
```
(where tmp is a directory in authorization-logic). Then, if the queries at the bottom are true, the CSVs in tmp will have one element and be empty otherwise.

As-is, with UserA's `privacyResolution` set to "stay_on" it works like the variant (mic stays on with UserC in the room). With the resolution set to "meet_requests", it works like the PRD scenario without the variant (mic turns off with UserC in room).

The main point of this PR is to make sure we have at least one way of doing both of these seemingly contradictory things:
- keep the core functionality that for any action to be done with data, the owner of the data must consent to it
- handle cases where requiring unanimous decisions won't work

(There might be other ways, but I wanted to help make sure we had at least one that runs).

### CODE REVIEW

File: src/analysis/souffle/examples/multimic/multimic-preferences.authlogic
Comment (aferr) [reply src/analysis/souffle/examples/multimic/multimic-preferences.authlogic:281]: We can once this is merged: [Implement may/will check #209
](https://github.com/google-research/raksha/pull/209#pullrequestreview-803727127)

File: src/analysis/souffle/examples/multimic/multimic-preferences.authlogic
Comment (aferr) [reply src/analysis/souffle/examples/multimic/multimic-preferences.authlogic:82]: Sorry this is just a typo, it should say `UserB`.

Actually, this statement that B is making about UserA will not be believed by `AmbientPrivacyPolicy`, because it only trusts users to state their own `setsPrivacyResolution` (via the rule on line 165). So if this typo happened in the wild, it wouldn't lead to a bad access decision.

File: src/analysis/souffle/examples/multimic/multimic-preferences.authlogic
Comment (aferr) [reply src/analysis/souffle/examples/multimic/multimic-preferences.authlogic:193]: Because [the implementation of the may/will check ](https://github.com/google-research/raksha/pull/209#pullrequestreview-803727127) works by checking if the original data owner says something `may` happen, the only way these rules could  lead to the application being accepted (in other words prove that the `will` part is allowed) is if the user delegated to the `"AmbientPrivacyPolicy"` (which in this example the users do -- for example on lines `47-52` is where UserA does this).

File: src/analysis/souffle/examples/multimic/multimic-preferences.authlogic
Comment (aferr) [reply src/analysis/souffle/examples/multimic/multimic-preferences.authlogic:242]: I'm assuming we can essentially:
  - include the whole authorization logic syntax in particle specs (I think you proposed parser combinators, which might help us do this).
  - when handles are referenced, our tool does the right thing and unfolds the paths properly whenever the particles are instantiated.

File: src/analysis/souffle/examples/multimic/multimic-preferences.authlogic
Comment (aferr) [reply src/analysis/souffle/examples/multimic/multimic-preferences.authlogic:281]: Or there might be a way to check these queries with tests without having them generated mechanically yet -- i figure if we like this idea, we might as well jump to having the whole end to end test.

File: src/analysis/souffle/examples/multimic/multimic-preferences.authlogic
Comment (aferr) [reply src/analysis/souffle/examples/multimic/multimic-preferences.authlogic]: Fixed the typo now.

**Comment by aferr** [comment]
@bgogul let me know if you would like to:
- merge this in
- wait until #209 is in and then get this tested against that
- just not merge this

**Comment by bgogul** [comment]
> @bgogul let me know if you would like to:
> 
> * merge this in
> * wait until [Implement may/will check #209](https://github.com/google-research/raksha/pull/209) is in and then get this tested against that
> * just not merge this

Let us merge this in after #209 so that this is tested as well.

File: src/analysis/souffle/examples/multimic/multimic-preferences.authlogic
Comment (markww) [inline src/analysis/souffle/examples/multimic/multimic-preferences.authlogic:15]: I recognize the value of this EULA pattern; the expression of the multimic policy in this file is much more auth-logic-idiomatic and rich than examples that do not use it. It does feel a little like this is manufacturing a centralized authority in a decentralized system; you could ask whether an EULA that is tied to buying the device offers users meaningful control over their privacy if there are only two real phone OS choices in the world.

I think you could potentially avoid the EULA-based approach by using the ownership tags that @bgogul introduced. This would involve the following steps:

1. Analyze the mic's subgraph in isolation, and from the combination of the mic's subgraph and UserA's `stay_on` policy, push to the edges of the subgraph the requirement that any owner that flows into the system allows for `userSpeech` to be downgraded to `textTranscript`.

2. Analyze the smartphone's graph in isolation and propagate ownership tags to the edges of the (presumably small) smartphone graph.

3. Upon connection, check the ownership tag against the mic's requirement.

I think that might serve our purpose without needing the EULA. That said, I see the value of the EULA pattern; perhaps we could develop both of these approaches as separate examples. I recognize additionally that the EULA pattern works right now, whereas the more static-analysis heavy approach requires more mechanism, so perhaps this should be the canonical example for a bit.

**Review by markww — APPROVED** [review]
My comment about "do we need an EULA" aside, this looks good. I'm fine with it being merged in as our primary multimic example.

File: src/analysis/souffle/examples/multimic/multimic-preferences.authlogic
Comment (aferr) [reply src/analysis/souffle/examples/multimic/multimic-preferences.authlogic:15]: > It does feel a little like this is manufacturing a centralized authority in a decentralized system; you could ask whether an EULA that is tied to buying the device offers users meaningful control over their privacy if there are only two real phone OS choices in the world.

I agree with this, actually. The whole point of the way policy checking works with a DIFC-ish taint analysis, authorization logic, and the "may/will" check is that absolutely everything that happens to data owned by some party (e.g. a user) can only be done with this data owner's consent. The PRD directly contrasts the way this policy checking normally works, so it is a pretty difficult case to handle. By default, this policy system would normally require every user to agree to everything that might happen with their data.

The mechanical part of this pattern just works by having all the data owners consent to having their preferences overridden under some situations (by doing a delegation). The "EULA" explanation was just my way of connecting this to something tied in reality.

File: src/analysis/souffle/examples/multimic/multimic-preferences.authlogic
Comment (aferr) [inline src/analysis/souffle/examples/multimic/multimic-preferences.authlogic:245]: @bgogul here was another way to express downgrading speech by all the users -- make a relation for the type of data (userSpeech) and add all the user speech tags to this relation. Populating the relation wasn't done yet, though, so the ownership thing is fine.

Comment (aferr) [reply src/analysis/souffle/examples/multimic/multimic-preferences.authlogic:245]: @markww not sure if you noticed this actually.

Comment (aferr) [reply src/analysis/souffle/examples/multimic/multimic-preferences.authlogic:245]: @markww not sure if you noticed this in the code -- you can downgrade all the data of a particular "type" by packing it into a relation for that type. I used this as a workaround for not having namespaces, which I used magically in the other PR before this one.

**Comment by aferr** [comment]
@bgogul also converted this one into an auth-logic test. PTAL.

**Pull request merged**
